File: src/main/java/com/github/joelws/release/tracker/service/artifact/operation/ReadArtifactServiceOperation.java
package com.github.joelws.release.tracker.service.artifact.operation;

import com.github.joelws.release.tracker.conversion.ArtifactDtoToArtifactAdapter;
import com.github.joelws.release.tracker.conversion.ArtifactToArtifactDtoAdapter;
import com.github.joelws.release.tracker.dto.artifact.ArtifactDto;
import com.github.joelws.release.tracker.entity.artifact.Artifact;
import com.github.joelws.release.tracker.handler.JsonResponse;
import com.github.joelws.release.tracker.service.ServiceHelper;
import com.github.joelws.release.tracker.service.ServiceOperation;
import com.github.joelws.release.tracker.service.artifact.execution.ReadArtifactServiceExecution;

import org.apache.log4j.Logger;

import javax.ws.rs.core.Response;

public class ReadArtifactServiceOperation extends ServiceOperation<String> {
    private static final Logger LOGGER = Logger.getLogger(ReadArtifactServiceOperation.class);

    private final ServiceHelper helper;

    private final ReadArtifactServiceExecution readArtifactServiceExecution;

    public ReadArtifactServiceOperation(final ServiceHelper helper,
                                        final ReadArtifactServiceExecution readArtifactServiceExecution) {
        this.helper = helper;
        this.readArtifactServiceExecution = readArtifactServiceExecution;
    }

    @Override
    protected Response delegate(final String param) {
        final ArtifactDtoToArtifactAdapter toArtifactDtoToArtifactAdapter = helper.getFactory().getImpl(ArtifactDtoToArtifactAdapter.class);

        final ArtifactDto artifactDto = helper.getJsonAdapter().getObjectFromJson(param, ArtifactDto.class);

        final Artifact result = readArtifactServiceExecution.execute(toArtifactDtoToArtifactAdapter.adapt(artifactDto));

        Response response;

        if (result != null) {
            final ArtifactToArtifactDtoAdapter toArtifactToArtifactDtoAdapter = helper.getFactory().getImpl(ArtifactToArtifactDtoAdapter.class);
            final ArtifactDto adaptedResult = toArtifactToArtifactDtoAdapter.adapt(result);
            response = helper.getRestResponseBuilder().build(200, adaptedResult);
        } else {
            response = helper.getRestResponseBuilder().build(new JsonResponse(404, "Artifact doesn't exist"));
        }
        return response;
    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/service/artifact/operation/ReadArtifactServiceOperation.kt
package com.github.joelws.release.tracker.service.artifact.operation

import com.github.joelws.release.tracker.conversion.ArtifactAdapter
import com.github.joelws.release.tracker.conversion.ArtifactModelAdapter
import com.github.joelws.release.tracker.entity.artifact.Artifact
import com.github.joelws.release.tracker.interfaces.Adapter
import com.github.joelws.release.tracker.model.artifact.ArtifactModel
import com.github.joelws.release.tracker.response.RestResponse.NotFound
import com.github.joelws.release.tracker.response.RestResponse.SuccessWithEntity
import com.github.joelws.release.tracker.response.build
import com.github.joelws.release.tracker.service.ServiceExecution
import com.github.joelws.release.tracker.service.ServiceHelper
import com.github.joelws.release.tracker.service.ServiceOperation
import javax.ws.rs.core.Response

open class ReadArtifactServiceOperation(private val helper: ServiceHelper,
                                        private val readArtifactServiceExecution: ServiceExecution<Artifact, Artifact?>) : ServiceOperation<String> {

    override fun delegate(param: String?): Response {

        @Suppress("UNCHECKED_CAST")
        val artifactModelAdapter: Adapter<ArtifactModel, Artifact> = helper.adapterFactory.getAdapter(ArtifactModelAdapter::class.java) as Adapter<ArtifactModel, Artifact>

        val artifactModel = helper.jsonAdapter.getObjectFromJson(param, ArtifactModel::class.java)

        val result = readArtifactServiceExecution.execute(artifactModelAdapter.adapt(artifactModel))

        val response = if (result != null) {

            @Suppress("UNCHECKED_CAST")
            val artifactAdapter: Adapter<Artifact, ArtifactModel> = helper.adapterFactory.getAdapter(ArtifactAdapter::class.java) as Adapter<Artifact, ArtifactModel>

            val adaptedResult = artifactAdapter.adapt(result)

            SuccessWithEntity(adaptedResult)
        } else {
            NotFound("Artifact doesn't exist")
        }
        return response.build()
    }
}
